packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,299 @@
-import path from "node:path";
-import { UserError } from "./errors";
+import { readdir, readFile, stat } from "node:fs/promises";
+import * as path from "node:path";
+import chalk from "chalk";
+import { getType } from "mime";
+import PQueue from "p-queue";
+import prettyBytes from "pretty-bytes";
+import { fetchResult } from "./cfetch";
+import { formatTime } from "./deploy/deploy";
+import { FatalError, UserError } from "./errors";
+import { logger, LOGGER_LEVELS } from "./logger";
+import {
+	BULK_UPLOAD_CONCURRENCY,
+	MAX_ASSET_COUNT,
+	MAX_ASSET_SIZE,
+	MAX_UPLOAD_ATTEMPTS,
+	MAX_UPLOAD_GATEWAY_ERRORS,
+} from "./pages/constants";
+import { hashFile } from "./pages/hash";
+import { APIError } from "./parse";
 import type { Config } from "./config";
 
+type AssetManifest = { [path: string]: { hash: string; size: number } };
+type InitializeAssetsResponse = {
+	// string of file hashes per bucket
+	buckets: string[][];
+	jwt: string;
+};
+
+export type UploadPayloadFile = {
+	name: string;
+	hash: string;
+	contents: string;
+	metadata: {
+		contentType: string;
+	};
+};
+
+type UploadResponse = {
+	jwt?: string;
+};
+
+export const syncAssets = async (
+	accountId: string | undefined,
+	scriptName: string,
+	assetDirectory: string | undefined,
+	dryRun: boolean | undefined
+): Promise<{
+	manifest: AssetManifest | undefined;
+	jwt: string | undefined;
+}> => {
+	if (assetDirectory === undefined) {
+		return { manifest: undefined, jwt: undefined };
+	}
+	if (dryRun) {
+		logger.log("(Note: doing a dry run, not uploading or deleting anything.)");
+		return { manifest: undefined, jwt: undefined };
+	}
+	assert(accountId, "Missing accountId");
+
+	const directory = path.resolve(assetDirectory);
+
+	// 1. generate asset manifest
+	const manifest = await walk(directory, {});
+
+	// 2. fetch buckets w/ hashes
+	const initializeAssetsResponse = await fetchResult<InitializeAssetsResponse>(
+		`/accounts/${accountId}/workers/scripts/${scriptName}/assets-upload-session`,
+		{
+			method: "POST",
+			body: JSON.stringify({ manifest: manifest }),
+		}
+	);
+
+	// if nothing to upload, return
+	if (initializeAssetsResponse.buckets.flat().length === 0) {
+		return { manifest, jwt: initializeAssetsResponse.jwt };
+	}
+
+	// 3. fill buckets and upload assets
+	const includedHashes = initializeAssetsResponse.buckets.flat();
+	const filteredFiles = Object.entries(manifest).filter((entry) =>
+		includedHashes.includes(entry[1].hash)
+	);
+	const queue = new PQueue({ concurrency: BULK_UPLOAD_CONCURRENCY });
+	let attempts = 0;
+	const start = Date.now();
+	let assetLogCount = 0;
+	let bucketUploadCount = 0;
+	let completionJwt = "";
+
+	for (const bucket of initializeAssetsResponse.buckets) {
+		attempts = 0;
+		let gatewayErrors = 0;
+		const doUpload = async (): Promise<UploadResponse> => {
+			// Populate the payload only when actually uploading (this is limited to 3 concurrent uploads at 50 MiB per bucket meaning we'd only load in a max of ~150 MiB)
```

Comment:
sanity check: do we need to dbl check these upload size limitations with EWC folks?